bdns/problem.go

```diff
@@ -15,6 +15,57 @@ type Error struct {
 	// Exactly one of rCode or underlying should be set.
 	underlying error
 	rCode      int
+
+	// Optional: If the resolver returned extended error information, it will be stored here.
+	// https://www.rfc-editor.org/rfc/rfc8914
+	extended *dns.EDNS0_EDE
+}
+
+// extendedDNSError returns non-nil if the input message contained an OPT RR
+// with an EDE option. https://www.rfc-editor.org/rfc/rfc8914.
+func extendedDNSError(msg *dns.Msg) *dns.EDNS0_EDE {
+	for _, extra := range msg.Extra {
+		if extra.Header().Rrtype != dns.TypeOPT {
+			continue
+		}
+
+		opt, ok := extra.(*dns.OPT)
+		if !ok {
+			continue
+		}
+
+		for _, opt := range opt.Option {
+			ede, ok := opt.(*dns.EDNS0_EDE)
+			if !ok {
+				continue
+			}
+			return ede
+		}
+	}
+	return nil
+}
+
+// wrapErr returns a non-nil error if err is non-nil or if resp.Rcode is not dns.RcodeSuccess.
+// The error includes appropriate details about the DNS query that failed.
+func wrapErr(queryType uint16, hostname string, resp *dns.Msg, err error) error {
```

Comment:
Looks like we don't have any test cases in bdns/dns_test.go that exercise non-nil return paths of wrapErr?